es/tx/tx.js

```diff
@@ -30,313 +30,137 @@ import { buildTx, calculateFee } from './builder'
-  // Build transaction using sdk (if nativeMode) or build on `AETERNITY NODE` side
-  return this.nativeMode
-    ? {
-      ...buildTx(R.merge(R.head(arguments), { nonce, ttl, fee, ctVersion, gasPrice }), TX_TYPE.contractCreate),
-      contractId: buildContractId(ownerId, nonce)
-    }
-    : this.api.postContractCreate(R.merge(R.head(arguments), { nonce, ttl, fee: parseInt(fee), gas: parseInt(gas), gasPrice, vmVersion: ctVersion.vmVersion, abiVersion: ctVersion.abiVersion }))
+function getTxBuilder (txType, fieldsMapper = a => a) {
+  return async function (options) {
+    let fields = await Promise.resolve(fieldsMapper.call(this, options))
+    // Calculate fee, get absolute ttl (ttl + height), get account nonce
+    const { fee, ttl, nonce } = await this.prepareTxParams(txType, fields)
+
+    fields = { ...fields, nonce, ttl, fee }
+    // Build transaction using sdk (if nativeMode) or build on `AETERNITY NODE` side
+    const { tx } = this.nativeMode
+      ? buildTx(fields, txType)
+      : await this.api[NODE_METHOD_NAME[txType]]({
+        ...fields,
+        ...fields.amount && { amount: parseInt(fields.amount) },
+        ...fields.fee && { fee: parseInt(fields.fee) },
+        ...fields.gas && { gas: parseInt(fields.gas) },
+        ...fields.initiatorAmountFinal && { initiatorAmountFinal: parseInt(fields.initiatorAmountFinal) },
+        ...fields.responderAmountFinal && { responderAmountFinal: parseInt(fields.responderAmountFinal) }
+      })
+
+    return tx
+  }
 }
 
-async function contractCallTx ({ callerId, contractId, abiVersion, amount, gas, gasPrice = MIN_GAS_PRICE, callData }) {
-  const ctVersion = this.getVmVersion(TX_TYPE.contractCall, R.head(arguments))
-  // Calculate fee, get absolute ttl (ttl + height), get account nonce
-  const { fee, ttl, nonce } = await this.prepareTxParams(TX_TYPE.contractCall, { senderId: callerId, ...R.head(arguments), gasPrice, abiVersion: ctVersion.abiVersion })
+const spendTx = getTxBuilder(
+  TX_TYPE.spend,
+  ({ payload = '', ...otherFields }) => ({ payload, ...otherFields })
+)
+
+const namePreclaimTx = getTxBuilder(TX_TYPE.namePreClaim)
+const nameClaimTx = getTxBuilder(TX_TYPE.nameClaim)
+const nameTransferTx = getTxBuilder(TX_TYPE.nameTransfer)
+const nameUpdateTx = getTxBuilder(TX_TYPE.nameUpdate)
+const nameRevokeTx = getTxBuilder(TX_TYPE.nameRevoke)
+
+const contractCreateTx = getTxBuilder(
+  TX_TYPE.contractCreate,
+  async function ({ ownerId, gasPrice = MIN_GAS_PRICE, ...otherFields }) {
+    const ctVersion = this.getVmVersion(TX_TYPE.contractCreate, arguments[0])
+    return ({
+      ownerId,
+      gasPrice,
+      ...otherFields,
+      ctVersion,
+      vmVersion: ctVersion.vmVersion,
+      abiVersion: ctVersion.abiVersion,
+      contractId: buildContractId(
+        ownerId,
+        (await this.prepareTxParams(TX_TYPE.oracleQuery, arguments[0])).nonce
```

Comment:
We calculate nonce twice for this tx???

Reply:
yes it is true, I can fix it if this PR would be accepted in general